src/runtime-rs/crates/resource/src/volume/block_volume.rs

```diff
@@ -35,76 +38,97 @@ impl BlockVolume {
         cid: &str,
         sid: &str,
     ) -> Result<Self> {
-        let fstat = stat::stat(m.source.as_str()).context(format!("stat {}", m.source))?;
-        info!(sl!(), "device stat: {:?}", fstat);
-        let mut options = HashMap::new();
-        if read_only {
-            options.insert("read_only".to_string(), "true".to_string());
-        }
-
-        let block_device_config = &mut BlockConfig {
-            major: stat::major(fstat.st_rdev) as i64,
-            minor: stat::minor(fstat.st_rdev) as i64,
-            ..Default::default()
+        let mnt_src: &str = &m.source;
+        // default block device fs type: ext4.
+        let mut blk_dev_fstype = DEFAULT_VOLUME_FS_TYPE.to_string();
+
+        let block_device_config = match m.r#type.as_str() {
+            KATA_MOUNT_BIND_TYPE => {
+                let fstat = stat::stat(mnt_src).context(format!("stat {}", m.source))?;
+
+                BlockConfig {
+                    major: stat::major(fstat.st_rdev) as i64,
+                    minor: stat::minor(fstat.st_rdev) as i64,
+                    ..Default::default()
+                }
+            }
+            KATA_DIRECT_VOLUME_TYPE => {
+                // get volume mountinfo from mountinfo.json
+                let v = volume_mount_info(mnt_src)
+                    .context("deserde information from mountinfo.json")?;
+                // check volume type
+                if v.volume_type != KATA_DIRECT_VOLUME_TYPE {
+                    return Err(anyhow!("volume type is invalid"));
+                }
+
+                let fstat = stat::stat(v.device.as_str())
+                    .with_context(|| format!("stat volume device file: {}", v.device.clone()))?;
+                if SFlag::from_bits_truncate(fstat.st_mode) != SFlag::S_IFREG
+                    && SFlag::from_bits_truncate(fstat.st_mode) != SFlag::S_IFBLK
+                {
+                    warn!(sl!(), "invalid volume device file: {}", v.device);
+                    return Err(anyhow!("create direct volume failed {:?}", "OK"));
```

Comment:
Is the message lost valid message `anyhow!("create direct volume failed {:?}", "OK")`?

Reply:
fixed